### PR TITLE
Fix wrong "ambiguous label" warning

### PR DESCRIPTION
**Pull request by Julow**
Fix https://github.com/ocaml/odoc/issues/760

- Fix the "ambigous label declaration" warning. The previous version was incorrectly comparing label names instead of the full identifiers.
  Now the backends must also handle the full identifiers (third point).

- Remove the warning about references to ambiguous labels.
  If the label is explicitly declared, a warning will be generated there. However, there is no warnings if the conflicting label is automatically generated. This extra warning was meant for covering the second case.
  A solution to this is to disallow references to generated labels.

- The code in the issue still pose problems in the latex backend.
  This is due to a "disambiguation" pass that suffixes the generated labels in case of ambiguity, which was written to fix the TOC but with the same wrong assumption as the other checks.
  I'm still investigating this but the solution might be different whether we allow references to generated labels or not.

- Some improvements to `Env` (remove an edge case that was used to support the point 2).

### Disallow references to auto-generated labels ?

Added in https://github.com/ocaml/odoc/issues/259

This has the advantage of simply removing the problem added by this PR (reference to ambiguous auto-generated label is broken or point to the wrong thing) and ensuring the disambiguation pass doesn't interfere with references.

Few documents might rely on them:
- Ocamldoc doesn't allow that.
- It was not possible to reference labels generated from headings with more than one word until Odoc 2.0.

### CODE REVIEW

**Comment by jonludlam** [comment]
This superseded #761 yes?

File: test/xref2/refs/refs.md [hunk 1]
```diff
@@ -998,7 +998,7 @@ Reference to 'Y' is ambiguous. Please specify its kind: constructor-Y, module-Y.
    Y)
 # resolve_ref "Everything_ambiguous_in_sig.t" (* Some kinds are missing: label, type subst (would be "type-") *) ;;
 File "<test>":
-Reference to 't' is ambiguous. Please specify its kind: field-t, module-type-t, type-t, val-t.
+Reference to 't' is ambiguous. Please specify its kind: field-t, module-type-t, type-t, val-t, val-t.
```
Comment (Julow) [inline test/xref2/refs/refs.md:1001]: The two `val-` actually correspond to a `external` and a `val`. This is a bug in the `Find` module but instead of fixing it, I'm investigating removing that module entirely and using `Env` instead.

File: src/xref2/env.ml [hunk 2]
```diff
@@ -80,97 +80,97 @@ type kind =
...
 type t = {
   linking : bool;
   (* True if this is a linking environment - if not,
      we only put in modules, module types, types, classes and class types *)
   id : int;
-  elts : Elements.t;
+  elts : ElementsByName.t;
+      (** Elements mapped by their name. Queried with {!find_by_name}. *)
+  ids : ElementsById.t;
+      (** Elements mapped by their identifier. Queried with {!find_by_id}. *)
```
Comment (Julow) [inline src/xref2/env.ml:173]: This is just a code improvement and not necessary. It was used by an other failed attempt at fixing the issue and I kept the commit.

File: test/xref2/refs/refs.md (hunk 1 above)
Comment (jonludlam) [reply test/xref2/refs/refs.md:1001]: Might be better to leave that change to 2.1.0?

**Comment by Julow** [comment]
> This supersedes #761 yes?

It does ! #761 was fixing an edge case that this PR removed.

File: src/xref2/env.ml [hunk 3]
```diff
@@ -80,97 +80,99 @@ type kind =
...
+  module IdMap = Identifier.Maps.Any
+
+  type t = Component.Element.any IdMap.t
+
+  let empty = IdMap.empty
+
+  let add identifier element t =
+    let check_for_duplicates () =
+      if IdMap.mem (identifier :> Identifier.t) t then
+        Format.eprintf "Duplicate found: %a\n%!" Component.Fmt.model_identifier
```
Comment (jonludlam) [inline src/xref2/env.ml:151]: We've lost the `assert` here

**Comment by jonludlam** [comment]
Once the assert is back in I think this is good to merge.

**Comment by Julow** [comment]
Before merging, what about the auto-generated labels ? Ambiguous references to these are no longer warned about.

**Comment by jonludlam** [comment]
Well we're certainly not removing it in this bugfix release :-) It'd be better if we did warn, but since the sitatution with this PR merged is enormously better than before I'm quite happy to go with what we've got here now.
